tests/signal/tst_signal.cpp

```diff
@@ -77,6 +79,128 @@ TEST_CASE("Signal connections")
         REQUIRE(lambdaCalled == true);
     }
 
+    SUBCASE("Disconnect Deferred Connection")
+    {
+        Signal<int> signal;
+        int val = 4;
+        auto evaluator = std::make_shared<ConnectionEvaluator>();
+
+        auto connection = signal.connectDeferred(evaluator, [&val](int value) {
+            val += value;
+        });
+
+        REQUIRE(connection.isActive());
+
+        signal.emit(4);
+
+        connection.disconnect();
```

Comment:
The reason I believe it is dangerous to have the previously emitted connection still be evaluated is:

Imagine that `val` was a `int *` allocated with `int* val = new int;`

I would expect this to then work:
```cpp
connection.disconnect();
delete val;
```

Then calling `evaluateDeferredConnections` would still reference val, as it's part of the connected closure.
This access would then be a dangling pointer and most likely SegFault, or write to random memory.

That's why I'm concerned about this behavior, you simply cannot guarantee that if you emitted a signal that references some resources those resources aren't referenced anymore after disconnecting.
Though that's exactly what disconnect is meant for in a non-deferred scenario.
The current behavior needs additional synchronization to ensure that both `disconnect` and `evaluateDeferredConnections` are called before any resources are freed.
Especially as those two calls are likely to happen in completely different threads the synchronization might be very difficult.

On the other hand I also get that it may be unintuitive that a signal emission doesn't actually execute a slot if it's not evaluated before it's disconnected.
Maybe @seanharmer or @lemirep have a preference on this, as they're using KDBindings quite extensively afaik.

Reply:
Thanks for the explanation! That does make sense. Now assuming there is no trick to extend the lifetime of closure object, seems like we have to safely synchronize it, or basically make sure that evaluation happen before the call to disconnect. 

I understand your point here, but I am concerned about how we used to understand signal, slot emit and disconnect. My intution while looking at code that contains structure like ` connectedDeferred -> emit -> disconnect -> evaluateDeferred` , I would like to see it as the classic implementation, means if signal is emitted before disconnection, it should be evaluated , no matter if the call to evaluation is after disconnecting the signal. But sadly that does have the problem with dangling reference. I would also love to see , views from @seanharmer and @lemirep 

> On the other hand I also get that it may be unintuitive that a signal emission doesn't actually execute a slot if it's not
> evaluated before it's disconnected.